src/lib/swapper/swappers/LifiSwapper/getTradeQuote/getTradeQuote.ts

```diff
@@ -131,81 +129,82 @@ export async function getTradeQuote(
     )
   }
 
-  // this corresponds to a "hop", so we could map the below code over selectedLifiRoute.steps to
-  // generate a multi-hop quote
-  const steps = await Promise.all(
-    selectedLifiRoute.steps.map(async lifiStep => {
-      // for the rate to be valid, both amounts must be converted to the same precision
-      const estimateRate = convertPrecision({
-        value: selectedLifiRoute.toAmountMin,
-        inputExponent: buyAsset.precision,
-        outputExponent: sellAsset.precision,
-      })
-        .dividedBy(bn(selectedLifiRoute.fromAmount))
-        .toString()
-
-      const protocolFees = transformLifiStepFeeData({
-        chainId,
-        lifiStep,
-        assets,
-      })
-
-      const buyAmountCryptoBaseUnit = bnOrZero(selectedLifiRoute.toAmountMin)
-      const intermediaryTransactionOutputs = getIntermediaryTransactionOutputs(assets, lifiStep)
-
-      const networkFeeCryptoBaseUnit = await getNetworkFeeCryptoBaseUnit({
-        accountNumber,
-        chainId,
-        lifiStep,
-        supportsEIP1559,
-        wallet,
-      })
-
-      return {
-        allowanceContract: lifiStep.estimate.approvalAddress,
-        accountNumber,
-        buyAmountBeforeFeesCryptoBaseUnit: buyAmountCryptoBaseUnit.toString(),
-        buyAsset,
-        intermediaryTransactionOutputs,
-        feeData: {
-          protocolFees,
-          networkFeeCryptoBaseUnit,
-        },
-        // TODO(woodenfurniture):  this step-level key should be a step-level value, rather than the top-level rate.
-        // might be better replaced by inputOutputRatio downstream
-        rate: estimateRate,
-        sellAmountIncludingProtocolFeesCryptoBaseUnit,
-        sellAsset,
-        sources: [
-          { name: `${selectedLifiRoute.steps[0].tool} (${SwapperName.LIFI})`, proportion: '1' },
-        ],
-      }
-    }),
+  return Ok(
+    await Promise.all(
```

Comment:
This could theoretically throw if one fails, we probably want to keep the previous monadic error-handling and only return :ok if this is actually successful?

Reply:
We actually don't want to return an error monad here, that is unless we want to rewrite to allow individual routes to be a monad result. Currently the assumption is that the whole quote request succeeds of fails such that the result is either an error monad (single), or an array of quotes

Reply:
The decision was essentially that if getting a quote failed at all, it's a fault in the swapper, and thus an error monad indictes a failure to fetcha quote.

We shouldn't try to make individual routes be monads because otherwise we would possibly end up showing several failed routes for a given swapper in the UI, when the user only cares that the swapper failed

Reply:
There might be valid in making the whole swapper quote request fail only if all routes throw. May we do this as a follow up task?

Reply:
> There might be valid in making the whole swapper quote request fail only if all routes throw. May we do this as a follow up task?

Thanks for the context 🙏🏽 . Agreed as it currently stands, we should either fail or return an array of quotes as they are *currently* tightly coupled (i.e if THOR fails, this means both streaming and regular quotes failed).

Let's tackle this as a follow-up indeed, as with the current flow and the quotes we're getting, this will work for us, but would be nice-to-have to decouple this whenever we have capacity

Reply:
ticket to resolve this https://github.com/shapeshift/web/issues/5110